File: rust_bindings/python/test/test_extractor.py
import os
import pytest
import sys
import twitter_text
import yaml

def test_extractor_ctor():
    extractor = twitter_text.Extractor()
    assert extractor is not None

def test_accessors():
    extractor = twitter_text.Extractor()

    assert extractor.get_extract_url_without_protocol() == True
    extractor.set_extract_url_without_protocol(False)
    assert extractor.get_extract_url_without_protocol() == False

def test_yaml():
    extractor = twitter_text.Extractor()
    with open (r"rust/conformance/tests/extract.yml") as file:
        testmap = yaml.load(file, Loader=yaml.FullLoader)

        assert len(testmap["tests"]["mentions"]) > 0
        for test in testmap["tests"]["mentions"]:
            mentions = extractor.extract_mentioned_screennames(test["text"])
            for index, s in enumerate(mentions):
                assert s == test["expected"][index]
        
        assert len(testmap["tests"]["mentions_with_indices"]) > 0
        for test in testmap["tests"]["mentions_with_indices"]:
            entities = extractor.extract_mentioned_screennames_with_indices(test["text"])
            assert len(test["expected"]) == len(entities)
            for index, entity in enumerate(entities):
                assert entity.value == test["expected"][index]["screen_name"]
                assert entity.start == test["expected"][index]["indices"][0]
                assert entity.end == test["expected"][index]["indices"][1]              

        assert len(testmap["tests"]["mentions_or_lists_with_indices"]) > 0
        for test in testmap["tests"]["mentions_or_lists_with_indices"]:
            entities = extractor.extract_mentions_or_lists_with_indices(test["text"])
            assert len(test["expected"]) == len(entities)
            for index, entity in enumerate(entities):
                assert entity.value == test["expected"][index]["screen_name"]
                assert entity.list_slug == test["expected"][index]["list_slug"]
                assert entity.start == test["expected"][index]["indices"][0]
                assert entity.end == test["expected"][index]["indices"][1]        

        assert len(testmap["tests"]["replies"]) > 0
        for test in testmap["tests"]["replies"]:
            entity = extractor.extract_reply_screenname(test["text"])
            if entity is not None:
                assert entity.value == test["expected"]
            else:
                assert test["expected"] is None

        assert len(testmap["tests"]["urls"]) > 0
        for test in testmap["tests"]["urls"]:
            urls = extractor.extract_urls(test["text"])
            for index, s in enumerate(urls):
                assert s == test["expected"][index]                

        assert len(testmap["tests"]["urls_with_indices"]) > 0
        for test in testmap["tests"]["urls_with_indices"]:
            entities = extractor.extract_urls_with_indices(test["text"])
            assert len(test["expected"]) == len(entities)
            for index, entity in enumerate(entities):
                assert entity.value == test["expected"][index]["url"]
                assert entity.start == test["expected"][index]["indices"][0]
                assert entity.end == test["expected"][index]["indices"][1]  

        assert len(testmap["tests"]["urls_with_directional_markers"]) > 0
        for test in testmap["tests"]["urls_with_directional_markers"]:
            entities = extractor.extract_urls_with_indices(test["text"])
            assert len(test["expected"]) == len(entities)
            for index, entity in enumerate(entities):
                assert entity.value == test["expected"][index]["url"]
                assert entity.start == test["expected"][index]["indices"][0]
                assert entity.end == test["expected"][index]["indices"][1]  

        assert len(testmap["tests"]["tco_urls_with_params"]) > 0
        for test in testmap["tests"]["tco_urls_with_params"]:
            entities = extractor.extract_urls_with_indices(test["text"])
            assert len(test["expected"]) == len(entities)
            for index, entity in enumerate(entities):
                assert entity.value == test["expected"][index]

        assert len(testmap["tests"]["hashtags"]) > 0
        for test in testmap["tests"]["hashtags"]:
            hashtags = extractor.extract_hashtags(test["text"])
            assert len(test["expected"]) == len(hashtags)
            for index, hashtag in enumerate(hashtags):
                assert hashtag == test["expected"][index]

        assert len(testmap["tests"]["hashtags_from_astral"]) > 0
        for test in testmap["tests"]["hashtags_from_astral"]:
            hashtags = extractor.extract_hashtags(test["text"])
            assert len(test["expected"]) == len(hashtags)
            for index, hashtag in enumerate(hashtags):
                assert hashtag == test["expected"][index]

        assert len(testmap["tests"]["hashtags_with_indices"]) > 0
        for test in testmap["tests"]["hashtags_with_indices"]:
            entities = extractor.extract_hashtags_with_indices(test["text"])
            assert len(test["expected"]) == len(entities)
            for index, entity in enumerate(entities):
                assert entity.value == test["expected"][index]["hashtag"]
                assert entity.start == test["expected"][index]["indices"][0]
                assert entity.end == test["expected"][index]["indices"][1]

        assert len(testmap["tests"]["cashtags"]) > 0
        for test in testmap["tests"]["cashtags"]:
            cashtags = extractor.extract_cashtags(test["text"])
            assert len(test["expected"]) == len(cashtags)
            for index, cashtag in enumerate(cashtags):
                assert cashtag == test["expected"][index]

        assert len(testmap["tests"]["cashtags_with_indices"]) > 0
        for test in testmap["tests"]["cashtags_with_indices"]:
            entities = extractor.extract_cashtags_with_indices(test["text"])
            assert len(test["expected"]) == len(entities)
            for index, entity in enumerate(entities):
                assert entity.value == test["expected"][index]["cashtag"]
                assert entity.start == test["expected"][index]["indices"][0]
                assert entity.end == test["expected"][index]["indices"][1]

def test_validating_extractor_ctor():
    extractor = twitter_text.ValidatingExtractor(twitter_text.TwitterTextConfiguration())
    assert extractor is not None

def test_validating_extractor_accessors():
    extractor = twitter_text.ValidatingExtractor(twitter_text.TwitterTextConfiguration())

    assert extractor.get_extract_url_without_protocol() == True
    extractor.set_extract_url_without_protocol(False)
    assert extractor.get_extract_url_without_protocol() == False

    assert extractor.get_normalize() == True
    extractor.set_normalize(False)
    assert extractor.get_normalize() == False

def test_validating_extractor_yaml():
    extractor = twitter_text.ValidatingExtractor(twitter_text.TwitterTextConfiguration())
    with open (r"rust/conformance/tests/extract.yml") as file:
        testmap = yaml.load(file, Loader=yaml.FullLoader)
        
        assert len(testmap["tests"]["mentions_with_indices"]) > 0
        for test in testmap["tests"]["mentions_with_indices"]:
            result = extractor.extract_mentioned_screennames_with_indices(test["text"])
            assert len(test["expected"]) == len(result.entities)
            for index, entity in enumerate(result.entities):
                assert entity.value == test["expected"][index]["screen_name"]
                assert entity.start == test["expected"][index]["indices"][0]
                assert entity.end == test["expected"][index]["indices"][1]              

        assert len(testmap["tests"]["mentions_or_lists_with_indices"]) > 0
        for test in testmap["tests"]["mentions_or_lists_with_indices"]:
            result = extractor.extract_mentions_or_lists_with_indices(test["text"])
            assert len(test["expected"]) == len(result.entities)
            for index, entity in enumerate(result.entities):
                assert entity.value == test["expected"][index]["screen_name"]
                assert entity.list_slug == test["expected"][index]["list_slug"]
                assert entity.start == test["expected"][index]["indices"][0]
                assert entity.end == test["expected"][index]["indices"][1]    

        assert len(testmap["tests"]["replies"]) > 0
        for test in testmap["tests"]["replies"]:
            result = extractor.extract_reply_screenname(test["text"])
            if result.entity is not None:
                assert result.entity.value == test["expected"]
            else:
                assert test["expected"] is None

        assert len(testmap["tests"]["urls_with_indices"]) > 0
        for test in testmap["tests"]["urls_with_indices"]:
            result = extractor.extract_urls_with_indices(test["text"])
            assert len(test["expected"]) == len(result.entities)
            for index, entity in enumerate(result.entities):
                assert entity.value == test["expected"][index]["url"]
                assert entity.start == test["expected"][index]["indices"][0]
                assert entity.end == test["expected"][index]["indices"][1]

        assert len(testmap["tests"]["urls_with_directional_markers"]) > 0
        for test in testmap["tests"]["urls_with_directional_markers"]:
            result = extractor.extract_urls_with_indices(test["text"])
            assert len(test["expected"]) == len(result.entities)
            for index, entity in enumerate(result.entities):
                assert entity.value == test["expected"][index]["url"]
                assert entity.start == test["expected"][index]["indices"][0]
                assert entity.end == test["expected"][index]["indices"][1]

        assert len(testmap["tests"]["hashtags_with_indices"]) > 0
        for test in testmap["tests"]["hashtags_with_indices"]:
            result = extractor.extract_hashtags_with_indices(test["text"])
            assert len(test["expected"]) == len(result.entities)
            for index, entity in enumerate(result.entities):
                assert entity.value == test["expected"][index]["hashtag"]
                assert entity.start == test["expected"][index]["indices"][0]
                assert entity.end == test["expected"][index]["indices"][1]

        assert len(testmap["tests"]["cashtags_with_indices"]) > 0
        for test in testmap["tests"]["cashtags_with_indices"]:
            result = extractor.extract_cashtags_with_indices(test["text"])
            assert len(test["expected"]) == len(result.entities)
            for index, entity in enumerate(result.entities):
                assert entity.value == test["expected"][index]["cashtag"]
                assert entity.start == test["expected"][index]["indices"][0]
                assert entity.end == test["expected"][index]["indices"][1]

if __name__ == "__main__":
    raise SystemExit(pytest.main([__file__]))

File: rust_bindings/java/src/test/java/app/grafica/twittertext/TestExtractor.java
package app.grafica.twittertext;

import static org.junit.Assert.*;
import org.junit.Test;

import java.io.IOException;
import java.util.List;
import java.util.Map;

public class TestExtractor {
  static {
    TwitterTextJNILib.load();
  }

  @Test
  public void testConstructor() {
    Extractor extractor = new Extractor();
    assertNotNull(extractor);
  }

  @Test
  public void testAccessors() {
    Extractor extractor = new Extractor();
    assertNotNull(extractor);

    assertEquals(extractor.getExtractUrlWithoutProtocol(), true);
    extractor.setExtractUrlWithoutProtocol(false);
    assertEquals(extractor.getExtractUrlWithoutProtocol(), false);
  }

  @Test
  public void testYaml() throws IOException {
    Extractor extractor = new Extractor();
    assertNotNull(extractor);

    String path = "rust/conformance/tests/extract.yml"; 
    List<Map> tests = Yaml.loadConformanceData(path, "mentions");
    assert(tests.size() > 0);
    for (Map test : tests) {
      assertEquals(test.get("expected"), extractor.extractMentionedScreennames(test.get("text").toString()));
    }

    tests = Yaml.loadConformanceData(path, "mentions_with_indices");
    assert(tests.size() > 0);
    for (Map test : tests) {
      List<Entity> actual = extractor.extractMentionedScreennamesWithIndices(test.get("text").toString());
      List<Entity> expected = Yaml.getExpectedEntities(test, "screen_name", null);
      assert(actual.size() > 0);
      assertEquals(actual.size(), expected.size());
      for (int i = 0; i < actual.size(); i++) {
        assertEquals(expected.get(i).getStart(), actual.get(i).getStart());
        assertEquals(expected.get(i).getEnd(), actual.get(i).getEnd());
        assertEquals(expected.get(i).getValue(), actual.get(i).getValue());
        assertEquals(expected.get(i).getListSlug(), actual.get(i).getListSlug());
      }
    }

    tests = Yaml.loadConformanceData(path, "mentions_or_lists_with_indices");
    assert(tests.size() > 0);
    for (Map test : tests) {
      List<Entity> actual = extractor.extractMentionsOrListsWithIndices(test.get("text").toString());
      List<Entity> expected = Yaml.getExpectedEntities(test, "screen_name", "list_slug");
      assert(actual.size() > 0);
      assertEquals(actual.size(), expected.size());
      for (int i = 0; i < actual.size(); i++) {
        assertEquals(expected.get(i).getStart(), actual.get(i).getStart());
        assertEquals(expected.get(i).getEnd(), actual.get(i).getEnd());
        assertEquals(expected.get(i).getValue(), actual.get(i).getValue());
        assertEquals(expected.get(i).getListSlug(), actual.get(i).getListSlug());
      }
    }
    
    tests = Yaml.loadConformanceData(path, "replies");
    assert(tests.size() > 0);
    for (Map test : tests) {
      Entity e = extractor.extractReplyScreenname(test.get("text").toString());
      assertEquals(test.get("expected"), e != null ? e.getValue() : null);
    }

    tests = Yaml.loadConformanceData(path, "urls");
    assert(tests.size() > 0);
    for (Map test : tests) {
      assertEquals(test.get("expected"), extractor.extractUrls(test.get("text").toString()));
    }

    tests = Yaml.loadConformanceData(path, "urls_with_indices");
    assert(tests.size() > 0);
    for (Map test : tests) {
      List<Entity> actual = extractor.extractUrlsWithIndices(test.get("text").toString());
      List<Entity> expected = Yaml.getExpectedEntities(test, "url", null);
      assert(actual.size() > 0);
      assertEquals(actual.size(), expected.size());
      for (int i = 0; i < actual.size(); i++) {
        assertEquals(expected.get(i).getStart(), actual.get(i).getStart());
        assertEquals(expected.get(i).getEnd(), actual.get(i).getEnd());
        assertEquals(expected.get(i).getValue(), actual.get(i).getValue());
        assertEquals(expected.get(i).getListSlug(), actual.get(i).getListSlug());
      }
    }

    tests = Yaml.loadConformanceData(path, "urls_with_directional_markers");
    assert(tests.size() > 0);
    for (Map test : tests) {
      List<Entity> actual = extractor.extractUrlsWithIndices(test.get("text").toString());
      List<Entity> expected = Yaml.getExpectedEntities(test, "url", null);
      assert(actual.size() > 0);
      assertEquals(actual.size(), expected.size());
      for (int i = 0; i < actual.size(); i++) {
        assertEquals(expected.get(i).getStart(), actual.get(i).getStart());
        assertEquals(expected.get(i).getEnd(), actual.get(i).getEnd());
        assertEquals(expected.get(i).getValue(), actual.get(i).getValue());
        assertEquals(expected.get(i).getListSlug(), actual.get(i).getListSlug());
      }
    }

    tests = Yaml.loadConformanceData(path, "tco_urls_with_params");
    assert(tests.size() > 0);
    for (Map test : tests) {
      assertEquals(test.get("expected"), extractor.extractUrls(test.get("text").toString()));
    }

    tests = Yaml.loadConformanceData(path, "hashtags");
    assert(tests.size() > 0);
    for (Map test : tests) {
      assertEquals(test.get("expected"), extractor.extractHashtags(test.get("text").toString()));
    }

    tests = Yaml.loadConformanceData(path, "hashtags_from_astral");
    assert(tests.size() > 0);
    for (Map test : tests) {
      assertEquals(test.get("expected"), extractor.extractHashtags(test.get("text").toString()));
    }

    tests = Yaml.loadConformanceData(path, "hashtags_with_indices");
    assert(tests.size() > 0);
    for (Map test : tests) {
      List<Entity> actual = extractor.extractHashtagsWithIndices(test.get("text").toString());
      List<Entity> expected = Yaml.getExpectedEntities(test, "hashtag", null);
      assert(actual.size() > 0);
      assertEquals(actual.size(), expected.size());
      for (int i = 0; i < actual.size(); i++) {
        assertEquals(expected.get(i).getStart(), actual.get(i).getStart());
        assertEquals(expected.get(i).getEnd(), actual.get(i).getEnd());
        assertEquals(expected.get(i).getValue(), actual.get(i).getValue());
        assertEquals(expected.get(i).getListSlug(), actual.get(i).getListSlug());
      }
    }

    tests = Yaml.loadConformanceData(path, "cashtags");
    assert(tests.size() > 0);
    for (Map test : tests) {
      assertEquals(test.get("expected"), extractor.extractCashtags(test.get("text").toString()));
    }

    tests = Yaml.loadConformanceData(path, "cashtags_with_indices");
    assert(tests.size() > 0);
    for (Map test : tests) {
      List<Entity> actual = extractor.extractCashtagsWithIndices(test.get("text").toString());
      List<Entity> expected = Yaml.getExpectedEntities(test, "cashtag", null);
      assert(actual.size() > 0);
      assertEquals(actual.size(), expected.size());
      for (int i = 0; i < actual.size(); i++) {
        assertEquals(expected.get(i).getStart(), actual.get(i).getStart());
        assertEquals(expected.get(i).getEnd(), actual.get(i).getEnd());
        assertEquals(expected.get(i).getValue(), actual.get(i).getValue());
        assertEquals(expected.get(i).getListSlug(), actual.get(i).getListSlug());
      }
    }
  }

  @Test
  public void tesValidatingConstructor() {
    ValidatingExtractor extractor = new ValidatingExtractor(TwitterTextConfiguration.configV3());
    assertNotNull(extractor);
  }


  @Test
  public void testValidatingAccessors() {
    ValidatingExtractor extractor = new ValidatingExtractor(TwitterTextConfiguration.configV3());
    assertNotNull(extractor);

    assertEquals(extractor.getNormalize(), true);
    extractor.setNormalize(false);
    assertEquals(extractor.getNormalize(), false);

    assertEquals(extractor.getExtractUrlWithoutProtocol(), true);
    extractor.setExtractUrlWithoutProtocol(false);
    assertEquals(extractor.getExtractUrlWithoutProtocol(), false);
  }
  @Test
  public void testValidatingYaml() throws IOException {
    ValidatingExtractor extractor = new ValidatingExtractor(new TwitterTextConfiguration());
    assertNotNull(extractor);

    String path = "rust/conformance/tests/extract.yml"; 
    List<Map> tests = Yaml.loadConformanceData(path, "mentions_with_indices");
    assert(tests.size() > 0);
    for (Map test : tests) {
      ExtractResult er = extractor.extractMentionedScreennamesWithIndices(test.get("text").toString());
      assert(er.getParseResults().getIsValid());
      List<Entity> actual = er.getEntities();
      List<Entity> expected = Yaml.getExpectedEntities(test, "screen_name", null);
      assert(actual.size() > 0);
      assertEquals(actual.size(), expected.size());
      for (int i = 0; i < actual.size(); i++) {
        assertEquals(expected.get(i).getStart(), actual.get(i).getStart());
        assertEquals(expected.get(i).getEnd(), actual.get(i).getEnd());
        assertEquals(expected.get(i).getValue(), actual.get(i).getValue());
        assertEquals(expected.get(i).getListSlug(), actual.get(i).getListSlug());
      }
    }
  
    tests = Yaml.loadConformanceData(path, "mentions_or_lists_with_indices");
    assert(tests.size() > 0);
    for (Map test : tests) {
      ExtractResult er = extractor.extractMentionsOrListsWithIndices(test.get("text").toString());
      assert(er.getParseResults().getIsValid());
      List<Entity> actual = er.getEntities();
      List<Entity> expected = Yaml.getExpectedEntities(test, "screen_name", "list_slug");
      assert(actual.size() > 0);
      assertEquals(actual.size(), expected.size());
      for (int i = 0; i < actual.size(); i++) {
        assertEquals(expected.get(i).getStart(), actual.get(i).getStart());
        assertEquals(expected.get(i).getEnd(), actual.get(i).getEnd());
        assertEquals(expected.get(i).getValue(), actual.get(i).getValue());
        assertEquals(expected.get(i).getListSlug(), actual.get(i).getListSlug());
      }
    }

    tests = Yaml.loadConformanceData(path, "replies");
    assert(tests.size() > 0);
    for (Map test : tests) {
      MentionResult mr = extractor.extractReplyScreenname(test.get("text").toString());
      Entity e = mr.getEntity();
      assertEquals(test.get("expected"), e != null ? e.getValue() : null);
    }

    tests = Yaml.loadConformanceData(path, "urls_with_indices");
    assert(tests.size() > 0);
    for (Map test : tests) {
      ExtractResult er = extractor.extractUrlsWithIndices(test.get("text").toString());
      assert(er.getParseResults().getIsValid());
      List<Entity> actual = er.getEntities();
      List<Entity> expected = Yaml.getExpectedEntities(test, "url", null);
      assert(actual.size() > 0);
      assertEquals(actual.size(), expected.size());
      for (int i = 0; i < actual.size(); i++) {
        assertEquals(expected.get(i).getStart(), actual.get(i).getStart());
        assertEquals(expected.get(i).getEnd(), actual.get(i).getEnd());
        assertEquals(expected.get(i).getValue(), actual.get(i).getValue());
        assertEquals(expected.get(i).getListSlug(), actual.get(i).getListSlug());
      }
    }

    tests = Yaml.loadConformanceData(path, "urls_with_directional_markers");
    assert(tests.size() > 0);
    for (Map test : tests) {
      ExtractResult er = extractor.extractUrlsWithIndices(test.get("text").toString());
      assert(er.getParseResults().getIsValid());
      List<Entity> actual = er.getEntities();
      List<Entity> expected = Yaml.getExpectedEntities(test, "url", null);
      assert(actual.size() > 0);
      assertEquals(actual.size(), expected.size());
      for (int i = 0; i < actual.size(); i++) {
        assertEquals(expected.get(i).getStart(), actual.get(i).getStart());
        assertEquals(expected.get(i).getEnd(), actual.get(i).getEnd());
        assertEquals(expected.get(i).getValue(), actual.get(i).getValue());
        assertEquals(expected.get(i).getListSlug(), actual.get(i).getListSlug());
      }
    }

    tests = Yaml.loadConformanceData(path, "hashtags_with_indices");
    assert(tests.size() > 0);
    for (Map test : tests) {
      ExtractResult er = extractor.extractHashtagsWithIndices(test.get("text").toString());
      assert(er.getParseResults().getIsValid());
      List<Entity> actual = er.getEntities();
      List<Entity> expected = Yaml.getExpectedEntities(test, "hashtag", null);
      assert(actual.size() > 0);
      assertEquals(actual.size(), expected.size());
      for (int i = 0; i < actual.size(); i++) {
        assertEquals(expected.get(i).getStart(), actual.get(i).getStart());
        assertEquals(expected.get(i).getEnd(), actual.get(i).getEnd());
        assertEquals(expected.get(i).getValue(), actual.get(i).getValue());
        assertEquals(expected.get(i).getListSlug(), actual.get(i).getListSlug());
      }
    }

    tests = Yaml.loadConformanceData(path, "cashtags_with_indices");
    assert(tests.size() > 0);
    for (Map test : tests) {
      ExtractResult er = extractor.extractCashtagsWithIndices(test.get("text").toString());
      assert(er.getParseResults().getIsValid());
      List<Entity> actual = er.getEntities();
      List<Entity> expected = Yaml.getExpectedEntities(test, "cashtag", null);
      assert(actual.size() > 0);
      assertEquals(actual.size(), expected.size());
      for (int i = 0; i < actual.size(); i++) {
        assertEquals(expected.get(i).getStart(), actual.get(i).getStart());
        assertEquals(expected.get(i).getEnd(), actual.get(i).getEnd());
        assertEquals(expected.get(i).getValue(), actual.get(i).getValue());
        assertEquals(expected.get(i).getListSlug(), actual.get(i).getListSlug());
      }
    }
  }
}